karapace/schema_backup.py

```diff
@@ -349,25 +349,72 @@ def _restore_backup_version_2(self, producer: KafkaProducer, fp: IO) -> None:
+            that it produces a string that is actually version 2 compatible. We anyway have to introduce a version 3,
+            and this public API can be fixed along with the introduction of it.
+
+        :param serialize: callback that encodes the consumer record into the target backup format.
+        :param poll_timeout: specifies the maximum time to wait for receiving records, if not records are received
+            within that time and the target offset has not been reached an exception is raised. Defaults to one second,
+            which is useful for testing and differs from the one minute CLI default.
+        :param overwrite: the output file if it exists.
+        :raises Exception: if consumption fails, concrete exception types are unknown, see Kafka implementation.
+        :raises FileExistsError: if ``overwrite`` is not ``True`` and the file already exists, or if the parent
+            directory of the file is not a directory.
+        :raises OSError: if writing fails or if the file already exists and is not actually a file.
+        :raises StaleConsumerError: if no records are received within the given ``poll_timeout`` and the target offset
+            has not been reached yet.
+        """
+        poll_timeout_ms = 1_000 if poll_timeout is None else poll_timeout // timedelta(milliseconds=1)
+        topic = self.topic_name
         with _writer(self.backup_location, overwrite=overwrite) as fp:
-            with _consumer(self.config, self.topic_name) as consumer:
-                LOG.info("Starting schema backup read for topic: %r", self.topic_name)
-
-                topic_fully_consumed = False
+            with _consumer(self.config, topic) as consumer:
+                (partition,) = consumer.partitions_for_topic(self.topic_name)
+                topic_partition = TopicPartition(self.topic_name, partition)
+                start_offset: int = consumer.beginning_offsets([topic_partition])[topic_partition]
+                end_offset: int = consumer.end_offsets([topic_partition])[topic_partition]
+                last_offset = start_offset
+                record_count = 0
 
                 fp.write(BACKUP_VERSION_2_MARKER)
-                while not topic_fully_consumed:
-                    raw_msg = consumer.poll(timeout_ms=self.timeout_ms, max_records=1000)
-                    topic_fully_consumed = len(raw_msg) == 0
-
-                    for _, messages in raw_msg.items():
-                        for message in messages:
-                            ser = export_func(key_bytes=message.key, value_bytes=message.value)
-                            if ser:
-                                fp.write(ser)
-
-                LOG.info("Schema export written to %r", "stdout" if fp is sys.stdout else self.backup_location)
+                if end_offset != 0:  # non-empty topic
+                    end_offset -= 1  # high watermark to actual end offset
```

Comment:
I would check if start offset and end offset are equal, this would be the rare case of a topic that is empty because of retention.